database/apikeys.go

```diff
@@ -0,0 +1,100 @@
+package database
+
+import (
+	"context"
+	"encoding/base64"
+	"time"
+
+	"gitlab.com/NebulousLabs/errors"
+	"gitlab.com/NebulousLabs/fastrand"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+/**
+API keys are authentication tokens generated by users. They do not expire, thus
+allowing users to use them for a long time and to embed them in apps and on
+machines. API keys can be revoked when they are no longer needed or if they get
+compromised. This is done by deleting them from this service.
+*/
+
+type (
+	// APIKey is a non-expiring authentication token generated on user demand.
+	APIKey struct {
+		ID     primitive.ObjectID `bson:"_id,omitempty" json:"id"`
+		UserID primitive.ObjectID `bson:"user_id" json:"-"`
+		// Key is a base64URL-encoded []byte with length PubKeySize
+		Key       string    `bson:"key" json:"key"`
+		CreatedAt time.Time `bson:"created_at" json:"createdAt"`
+	}
+)
+
+// APIKeyCreate creates a new API key.
+func (db *DB) APIKeyCreate(ctx context.Context, user User) (*APIKey, error) {
+	if user.ID.IsZero() {
+		return nil, errors.New("invalid user")
+	}
+	ak := APIKey{
+		UserID:    user.ID,
+		Key:       base64.URLEncoding.EncodeToString(fastrand.Bytes(PubKeySize)),
+		CreatedAt: time.Now().UTC(),
+	}
+	ior, err := db.staticAPIKeys.InsertOne(ctx, ak)
+	if err != nil {
+		return nil, err
+	}
+	ak.ID = ior.InsertedID.(primitive.ObjectID)
+	return &ak, nil
+}
+
+// APIKeyFetch fetches an API key from the DB.
+func (db *DB) APIKeyFetch(ctx context.Context, user User, id primitive.ObjectID) (*APIKey, error) {
```

Comment:
is this function used? the fact that it expects a `primitive.ObjectID` is a bit of a flag for me, the object id should not be exposed right? so how can we provide it to fetch an API key and where is it used

Reply:
Ha! No, it's not. I had an endpoint that fetched a single API key before (and the WIP version used `ObjectID`) but I removed it. I've forgotten to clean up. Thanks! :)

Reply:
Agreed. Even if it is used we should avoid weird internal types.